Library/Homebrew/cmd/commands.rb

```diff
@@ -8,51 +8,70 @@ module Homebrew
   module_function
 
   def commands
-    if ARGV.include? "--quiet"
-      cmds = internal_commands + external_commands
-      cmds += internal_developer_commands
-      cmds += HOMEBREW_INTERNAL_COMMAND_ALIASES.keys if ARGV.include? "--include-aliases"
-      puts Formatter.columns(cmds.sort)
-    else
-      # Find commands in Homebrew/cmd
-      puts "Built-in commands"
-      puts Formatter.columns(internal_commands)
-
-      # Find commands in Homebrew/dev-cmd
-      puts
-      puts "Built-in developer commands"
-      puts Formatter.columns(internal_developer_commands)
-
-      # Find commands in the path
-      unless (exts = external_commands).empty?
-        puts
-        puts "External commands"
-        puts Formatter.columns(exts)
+    # TODO: Goal is to not require this function at all.
+    command = CommandsCommand.new
+    command.check_for_errors
+    command.commands
+  end
+
+  class CommandsCommand < Command
+    def initialize
+      options do
+        cmd_name "commands"
+        desc "Show a list of built-in and external commands."
+        option "quiet", desc: "list only the names of commands without the header." do
+          option "include-aliases", desc: "the aliases of internal commands will be included."
```

Comment:
Sorry, shouldn't have said "works," meant readable. We can definitely make nested dependencies and exclusions work (they're subsets of basic problems), but expressing them in a DSL can quickly become really ugly:

```ruby
option "foo", desc: "foo" do
  option "bar", desc: "bar" do
    option "baz", desc: "baz" do

    end
  end
end
```

Contrast that with:

```ruby
option "foo", desc: "foo"
option "bar", desc: "bar"
option "baz", desc: "baz"
```

In the former, passing `--bar` or `--baz` without `--foo` is an error. In the latter `--bar` and `--baz` are just ignored, since they're "legal" options but simply have no meaning without `--foo`.

The former is more strictly correct in terms of catching unused options, but the unused options that we're *most* interested in catching are the *completely* wrong ones like typos and the latter does that just fine (and with no scoping or additional DSL code to think about).